internal/maintenance/job_scheduler.go

```diff
@@ -83,15 +91,28 @@ type JobScheduler struct {
 
 	config *JobSchedulerConfig
 	exec   riverdriver.Executor
+
+	// Circuit breaker that tracks consecutive timeout failures from the
+	// scheduling query. The query starts by using the full/default batch size,
+	// but after this breaker trips (after N consecutive timeouts occur in a
+	// row), it switches to a smaller batch. We assume that a database that's
+	// degraded is likely to stay degraded over a longer term, so after the
+	// circuit breaks, it stays broken until the program is restarted.
+	reducedBatchSizeBreaker *circuitbreaker.CircuitBreaker
 }
 
 func NewJobScheduler(archetype *baseservice.Archetype, config *JobSchedulerConfig, exec riverdriver.Executor) *JobScheduler {
 	return baseservice.Init(archetype, &JobScheduler{
+		reducedBatchSizeBreaker: circuitbreaker.NewCircuitBreaker(&circuitbreaker.CircuitBreakerOptions{
+			Limit:  3,
+			Window: 10 * time.Minute,
+		}),
 		config: (&JobSchedulerConfig{
-			Interval:     cmp.Or(config.Interval, JobSchedulerIntervalDefault),
-			Limit:        cmp.Or(config.Limit, JobSchedulerLimitDefault),
-			NotifyInsert: config.NotifyInsert,
-			Schema:       config.Schema,
+			BatchSizeDefault: cmp.Or(config.BatchSizeDefault, JobSchedulerBatchSizeDefault),
```

Comment:
It could also be that 10k is too much as a default for typical systems, but we can still change that later